server/auth/types/openid/routes.ts

```diff
@@ -0,0 +1,213 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+import {
+  IRouter,
+  SessionStorageFactory,
+  CoreSetup,
+} from '../../../../../../src/core/server';
+import { SecuritySessionCookie } from '../../../session/security_cookie';
+import { SecurityPluginConfigType } from '../../..';
+import { schema } from '@kbn/config-schema';
+import { randomString } from '@hapi/cryptiles';
+import { parse, stringify } from 'querystring';
+import { OpenIdAuthConfig } from './openid_auth';
+import wreck from '@hapi/wreck';
+import { SecurityClient } from '../../../backend/opendistro_security_client';
+
+export class OpenIdAuthRoutes {
+  constructor(
+    private readonly router: IRouter,
+    private readonly config: SecurityPluginConfigType,
+    private readonly sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    private readonly openIdAuthConfig: OpenIdAuthConfig,
+    private readonly securityClient: SecurityClient,
+    private readonly core: CoreSetup
+  ) {}
+
+  public setupRoutes() {
+    this.router.get(
+      {
+        path: `/auth/openid/login`,
+        validate: {
+          query: schema.object({
+            code: schema.maybe(schema.string()),
+            nextUrl: schema.maybe(schema.string()),
+            state: schema.maybe(schema.string()),
+            refresh: schema.maybe(schema.string()),
+          }),
+        },
+        options: {
+          authRequired: false,
+        },
+      },
+      async (context, request, response) => {
+        // implementation refers to https://github.com/hapijs/bell/blob/master/lib/oauth.js
+
+        // Sign-in initialization
+        if (!request.query.code) {
+          const nonce = randomString(22);
+          const query: any = {
+            client_id: this.config.openid?.client_id,
+            response_type: 'code',
+            redirect_uri: `${this.getBaseRedirectUrl()}`, // TODO: generate actual domain name
```

Comment:
Just curious what the todo means. Is the domain same as AES domain?

Reply:
removed, that was a TODO while I do implementation/testing, it is addressed already